backend/inspirehep/pidstore/providers/texkey.py

```diff
@@ -0,0 +1,174 @@
+            )
+        raise TexkeyCannotGenerateFirstPart
+
+    @classmethod
+    def build_texkey_second_part(cls, data):
+        date = get_literature_earliest_date(data)
+        if date:
+            return date.split("-", 1)[0]
+        raise TexkeyCannotGenerateSecondPart
+
+    @classmethod
+    def get_texkey_with_random_part(cls, texkey):
+        retry_count = current_app.config.get("PIDSTORE_TEXKEY_MAX_RETRY_COUNT", 5)
+        size = current_app.config.get("PIDSTORE_TEXKEY_RANDOM_PART_SIZE", 3)
+        used_texkeys = cls.query_all_texkeys_starting_with(texkey)
+        for _ in range(retry_count):
+            random_part = "".join(
+                random.choices(string.ascii_lowercase + string.digits, k=size)
+            )
+            pid_value = f"{texkey}{random_part}"
+            if pid_value not in used_texkeys:
+                return pid_value
+        raise CannotGenerateUniqueTexKey
+
+    @classmethod
+    def query_pid_for_texkey(cls, texkey):
+        texkey = (
+            PersistentIdentifier.query.filter(PersistentIdentifier.pid_value == texkey)
+            .filter(PersistentIdentifier.pid_type == cls.pid_type)
+            .one_or_none()
+        )
+        return texkey
+
+    @classmethod
+    def query_all_texkeys_starting_with(cls, texkey):
+        texkeys = (
+            PersistentIdentifier.query.with_for_update()
+            .with_entities(PersistentIdentifier.pid_value)
+            .filter(PersistentIdentifier.pid_type == cls.pid_type)
+            .filter(PersistentIdentifier.pid_value.startswith(texkey))
+            .all()
+        )
+        texkeys = set(flatten_list(texkeys))
+        return texkeys
+
+    @classmethod
+    def is_texkey_assigned_correctly(cls, texkey, uuid):
+        texkey = cls.query_pid_for_texkey(texkey)
+        if (
+            texkey
+            and texkey.object_uuid == uuid
+            and texkey.status == PIDStatus.REGISTERED
+        ):
+            return True
+        elif texkey:
+            raise PIDAlreadyExistsError(texkey.pid_type, texkey.pid_value)
+        return False
+
+    @classmethod
+    def create_or_update_texkey_in_pidstore(cls, texkey, uuid):
+        texkey = cls.query_pid_for_texkey(texkey)
+        if texkey:
+            if texkey.object_uuid != uuid and texkey.status == PIDStatus.DELETED:
+                texkey.object_uuid = uuid
+                texkey.status = PIDStatus.REGISTERED
+                texkey.save()
+            elif texkey.object_uuid != uuid and texkey.status == PIDStatus.REGISTERED:
+                raise PIDAlreadyExistsError(cls.pid_type, texkey)
+            elif texkey.object_uuid == uuid and texkey.status == PIDStatus.DELETED:
+                texkey.status = PIDStatus.REGISTERED
+                texkey.save()
+        return texkey
+
+    @classmethod
+    def create(
+        cls, object_type=None, object_uuid=None, data=None, pid_value=None, **kwargs
+    ):
+        if pid_value:
```

Comment:
I know you didn't invent this, but I just want to comment that the API here is really terrible: a function with a top-level `if` based on the presence of some arg is really two different functions.